### DISCUSSION TITLE
Realtime communication between a lot of clients. Websockets only

### DISCUSSION THREAD

**Discussion by hmatthieu** [trpc/trpc#6160] [RFC / Ideas]
Good morning tRPC community,

First off, I just want to say I’m a big fan of tRPC—massive thanks to the maintainers for all your work on this codebase 🙏. It’s been an amazing tool, and I’m excited to see how far it can go.

I’d like to start a discussion around optimizing the tRPC network layer for high-throughput scenarios, specifically focusing on uWebSocket integration and pubsub capabilities. I’ll briefly share my context to clarify the use case, but the focus remains on tRPC itself.

**Context**  
I’m prototyping an MMO web game where many users need to exchange data in real-time. tRPC has been my choice due to its type-safety and flexibility. I’ve evaluated alternatives (from lower-level solutions like protobuf to higher-level ones like Colyseus), but this discussion isn’t about switching tools; rather, it’s about pushing tRPC to its limits. However, if this use case seems misaligned with tRPC's capabilities, feel free to let me know before we dive in.

**The Problem**  
Imagine 100 users on a single server, all sharing real-time updates (movements, actions, etc.). Each client sends data to the server using mutations and listens for updates through subscriptions. Here’s a breakdown of a typical operation:
1. A client sends their new position to the server via a mutation.
2. The server verifies and routes this data through a server-side bus.
3. The server broadcasts the data to other clients via subscriptions.

From testing, step 3 is the most CPU-intensive due to the serialization and transmission over numerous WebSocket connections. For 100 users continuously updating, this can mean up to 10,000 server-to-client messages every few milliseconds.

**Potential Solutions**  
Using tRPC’s default subscription mechanism (based on `ws`), serialization and data dispatch account for much of the load. Switching to uWebSocket can nearly double the throughput, and eliminating repetitive serialization offers even more gains. Here’s how:
- If each client receives identical data, we can serialize once and broadcast the same message to all, leveraging uWebSocket’s pubsub to bypass Node.js's broadcasting.

However, here we encounter a limitation in tRPC:
tRPC's message IDs, used for routing responses, are generated per request, complicating broadcast scenarios where we’d want a known ID at publish time. 

I’ve experimented with some workarounds that surprisingly don’t require internal tRPC changes:
- **Create a uWebSocket adapter** (since [romanzy313/trpc-uwebsockets](https://github.com/romanzy313/trpc-uwebsockets) isn't compatible with the latest tRPC versions).
- **Use a custom link** to override the generated message ID with a predictable one.

The only caveat here is ID consistency; either we use a unique string ID per message or a high number (close to `MAX_SAFE_INTEGER`) to avoid conflicts with tRPC's IDs.

**Implementation Snapshot**  
Here’s how the setup looks in code. Note that the client code remains identical, maintaining a seamless experience:

_Server-side setup for pubsub:_
```typescript
mySubscription: procedure.subscription(() =>
  pubsub.subscribe<IMessage>("myRouter.mySubscription"),
),
```

_Publishing to clients:_
```typescript
pubsub.publish("myRouter.mySubscription", /* IMessage object */);
```

**Final Thoughts**  
- Could a native pubsub mechanism be a valuable addition to tRPC (e.g., a `.pubsub(() => topic)` procedure)? While it would simplify this setup, it could broaden tRPC's scope.
- Is tRPC suitable for real-time data-heavy applications like this? I believe it is, as tRPC’s strong type-safety aligns well with our needs. The goal here is simply to maximize throughput.

Thank you in advance for your insights!

**Comment by hmatthieu** [comment]
Might be related to https://github.com/trpc/trpc/issues/6109 ?

**Reply by KATT** [reply]
Would you wanna help?

**Reply by hmatthieu** [reply]
I’d love to help on this issue, but I don’t have the bandwidth right now. If it aligns with my current scope of work, I'll definitely pitch in.

**Reply by KATT** [reply]
If you're happy to help the library, we're happy to help you use the library

I'd love tRPC to cater for high-throughput scenarios, but it's not what any of us in the core team use it for, so we won't spend time building this - we're happy to give you a lot of guiding around the codebase though if you want to contribute. An alternative is a financial thing where you pay us to build the features you need.

**Reply by hmatthieu** [reply]
> I'd love tRPC to cater for high-throughput scenarios

That's all I wanted to hear 😄 

On a more serious note, through this post, I'm looking to connect with people facing similar scenarios to understand how I might best contribute to the project—whether through internal changes, adapters, an external library, or… nothing at all if I'm alone on this topic haha.
